Use a TypeVar for vmap/pmap/jit function types, not just their return values.

aslanides opened the pull request:
This has the benefit of preserving both call and return signatures.

Note that apparently at least two types must be passed as constraints to TypeVar, so we have the slightly odd and redundant Callable, Callable[..., Any].

aslanides commented:
Hmm, seems like `mypy` has a problem with this:


```
$ mypy jax
> jax/image/scale.py:222: error: "object" not callable
> jax/random.py:433: error: "object" not callable
```

Both of these errors seem to originate at applications of jax.jit via functools.partial ... although they are far from the only instances in the codebase and I can't seem to see anything distinctive about them.

shoyer commented on jax/api.py:
```diff
@@ -68,16 +68,14 @@
 # Should this raise any type errors for the tracing code in future, we can disable
 # type checking in parts of the tracing code, or remove these annotations.
-T = TypeVar("T")
+F = TypeVar("F", Callable, Callable[..., Any])
```
It looks like the [recommendation from mypy](https://mypy.readthedocs.io/en/stable/generics.html#declaring-decorators) is to use `bound`, e.g.,:
```suggestion
F = TypeVar("F", bound=Callable[..., Any])
```

I'm not entirely sure why that would be any different, but it might be worth a try...

aslanides commented:
Presubmit failure seems like a flaky doc test.

aslanides commented:
Failing test(s) seem to be due to unrelated flakes -- I think we're safe to merge.

jekbradbury commented:
This change causes pytype to fail on certain downstream code patterns. For example, [this code](https://github.com/deepmind/dm-haiku/blob/master/examples/impala/learner.py#L40):
```python
mean_per_batch = jax.vmap(rlax.policy_gradient_loss, in_axes=1)(logits, *args)
```
fails to typecheck, with
```
File "[...]/haiku/examples/impala/learner.py", line 40, in policy_gradient_loss: Missing parameter 'adv_t' in call to function rlax.policy_gradient_loss [missing-parameter]
         Expected: (logits_t, a_t, adv_t, w_t)
  Actually passed: (logits_t, a_t)
```

This case might be a pytype bug, since `policy_gradient_loss` is in fact called with the four arguments it expects.

trevorcai commented:
Happy to fix Haiku in the above scenario (as it's a pretty straightforward improvement over my lazy code), but do want to register that `*args` is useful in some cases and it'd be a pity to lose this.

jekbradbury commented:
There are enough distinct failures in downstream code resulting from `*args` that I don't think we should merge this unless/until we can fix that (either in this PR, or in pytype).

aslanides commented:
Thanks -- I've raised a bug with the PyType team: b/174530851.

aslanides commented:
Update on this: the pytype team have now resolved the issue with unpacking *args, and I have fixed other breakages in downstream internal code. PTAL @jekbradbury :)

aslanides committed:
Improve type safety for various function transformations.

This change expresses the fact that function signatures (both arguments and return types) are
invariant under {jit, vmap, pmap} transformations.
